### PR TITLE
WIP: Virtualbox Additions

### PR DESCRIPTION
**Pull request by wattoc**
I need to take a break from this one so sharing in case someone else wants to pick it up or help - tested on x64, needs testing on x86.  It may not build at all on x86 which is probably some sort of compiler values that need tweaking.

Enabled: video, mouse, shared folders, tray, console apps (service and control.)

Notes:

Tray:
Crashes on Quit or during shutdown if Quit before shutdown.
NB. you need to run this to enable shared clipboard and the video resizing

Shared folders:
Seems to work generally although Tracker is not updating after deletes, renames, etc.  tried messing about with put nodes and notifications but kind of stuck here.  There's an issue with deleting directories too, the VirtualBox side claims they are always in use - may be a Close or something I'm missing during directory walk or lookup or may even be a VirtualBox bug.

Video:
Seems to work fine although limited testing - it allows you to resize the Virtualbox window and have the display dynamically resize.  You need the tray running for that to work though.

Mouse:
I'm not sure it really does anything of value, but it doesn't crash - I'm thinking it should enable/disable mouse capture when you cross the Virtualbox display boundaries but kind of run out of steam at this point.

Clipboard:
Seems to work fine although only text seems to work so far in my testing.

There's plenty of logging, etc. that I've added trying to diagnose stuff, so that'd need cleaned up.  I've left the patch unflattened so others can see the progression of changes/revert stuff they think is superfluous.

Need to get back to some app type things for a bit as drivers are not fun.

### CODE REVIEW

**Comment by Begasus** [comment]
First error when trying to build on x86_gcc2:
```
Config.kmk:177: *** You must update kBuild! Requires kBuild binaries 3074 or later, kmk says 2656 (0.1.9998-r2656).  Stop
```

**Comment by wattoc** [comment]
It may not be able to build at all on GCC2.  Main problem with that would be the build for the Tray, which AFAIK needs to be GCC2 to actually work on x86.

Probably it's better to replace the tray app with a background app instead.

**Comment by Begasus** [comment]
> It may not be able to build at all on GCC2

It's using gcc8 (building for secondary architecture), in Terminal it seems to be OK (build starts when I use the commandline from the recipe)?

**Comment by Begasus** [comment]
Error when trying to build from Terminal (gcc8):

```
kBuild: Pass - DLLs
kBuild: Pass - Programs
kBuild: Compiling bldRTLdrCheckImports - /Opslag/haikuports/app-emulation/virtualbox-guest-additions/work-x86-6.1.26/sources/VirtualBox-6.1.26/src/VBox/Runtime/tools/RTLdrCheckImports.cpp
kBuild: Compiling VBoxControl - /Opslag/haikuports/app-emulation/virtualbox-guest-additions/work-x86-6.1.26/sources/VirtualBox-6.1.26/src/VBox/Additions/common/VBoxControl/VBoxControl.cpp
kBuild: Compiling VBoxService - /Opslag/haikuports/app-emulation/virtualbox-guest-additions/work-x86-6.1.26/sources/VirtualBox-6.1.26/src/VBox/Additions/common/VBoxService/VBoxService.cpp
kBuild: Compiling VBoxService - /Opslag/haikuports/app-emulation/virtualbox-guest-additions/work-x86-6.1.26/sources/VirtualBox-6.1.26/src/VBox/Additions/common/VBoxService/VBoxServiceUtils.cpp
kBuild: Compiling VBoxService - /Opslag/haikuports/app-emulation/virtualbox-guest-additions/work-x86-6.1.26/sources/VirtualBox-6.1.26/src/VBox/Additions/common/VBoxService/VBoxServiceStats.cpp
kBuild: Compiling VBoxService - /Opslag/haikuports/app-emulation/virtualbox-guest-additions/work-x86-6.1.26/sources/VirtualBox-6.1.26/src/VBox/Additions/common/VBoxService/VBoxServiceTimeSync.cpp
kBuild: Compiling VBoxService - /Opslag/haikuports/app-emulation/virtualbox-guest-additions/work-x86-6.1.26/sources/VirtualBox-6.1.26/src/VBox/Additions/common/VBoxService/VBoxServiceToolBox.cpp
kBuild: Compiling VBoxService - /Opslag/haikuports/app-emulation/virtualbox-guest-additions/work-x86-6.1.26/sources/VirtualBox-6.1.26/src/VBox/Additions/common/VBoxService/VBoxServiceBalloon.cpp
kBuild: Compiling VBoxService - /Opslag/haikuports/app-emulation/virtualbox-guest-additions/work-x86-6.1.26/sources/VirtualBox-6.1.26/src/VBox/Additions/common/VBoxService/VBoxServiceVMInfo.cpp
kBuild: Compiling VBoxService - /Opslag/haikuports/app-emulation/virtualbox-guest-additions/work-x86-6.1.26/sources/VirtualBox-6.1.26/src/VBox/Additions/common/VBoxService/VBoxServicePropCache.cpp
The failing command:
@g++ -c -O2 -g -fno-exceptions  -m32 -I/Opslag/haikuports/app-emulation/virtualbox-guest-additions/work-x86-6.1.26/sources/VirtualBox-6.1.26/src/VBox/Additions/haiku/include -I/Opslag/haikuports/app-emulation/virtualbox-guest-additions/work-x86-6.1.26/sources/VirtualBox-6.1.26/out/haiku.x86/strict/obj/VBoxMouseFilter/dtrace -I/Opslag/haikuports/app-emulation/virtualbox-guest-additions/work-x86-6.1.26/sources/VirtualBox-6.1.26/include -I/Opslag/haikuports/app-emulation/virtualbox-guest-additions/work-x86-6.1.26/sources/VirtualBox-6.1.26/out/haiku.x86/strict -DVBOX -DVBOX_OSE -DVBOX_WITH_64_BITS_GUESTS -DRT_STRICT -DVBOX_STRICT -DRT_OS_HAIKU -DRT_ARCH_X86 -D__X86__ -DVBOX_WITH_DEBUGGER -DRT_LOCK_STRICT -DRT_LOCK_STRICT_ORDER -DRTPATH_APP_PRIVATE=\"/boot/system/apps/VirtualBox\" -DRTPATH_APP_PRIVATE_ARCH=\"/boot/system/apps/VirtualBox\" -DRTPATH_SHARED_LIBS=\"/boot/system/apps/VirtualBox\" -DRTPATH_APP_DOCS=\"/boot/system/apps/VirtualBox\" -DIN_RING3 -DIN_GUEST -DIN_GUEST_R3 -DIN_RT_R3 -DGC_ARCH_BITS=64 -DRT_NEED_NEW_AND_DELETE -DVBOX_WITH_HGCM -DLOG_TO_BACKDOOR -DLOG_ENABLED -DHC_ARCH_BITS=32 -Wp,-MD,/Opslag/haikuports/app-emulation/virtualbox-guest-additions/work-x86-6.1.26/sources/VirtualBox-6.1.26/out/haiku.x86/strict/obj/VBoxMouseFilter/VBoxMouseFilter.o.dep -Wp,-MT,/Opslag/haikuports/app-emulation/virtualbox-guest-additions/work-x86-6.1.26/sources/VirtualBox-6.1.26/out/haiku.x86/strict/obj/VBoxMouseFilter/VBoxMouseFilter.o -Wp,-MP -o /Opslag/haikuports/app-emulation/virtualbox-guest-additions/work-x86-6.1.26/sources/VirtualBox-6.1.26/out/haiku.x86/strict/obj/VBoxMouseFilter/VBoxMouseFilter.o /Opslag/haikuports/app-emulation/virtualbox-guest-additions/work-x86-6.1.26/sources/VirtualBox-6.1.26/src/VBox/Additions/haiku/VBoxMouse/VBoxMouseFilter.cpp
kBuild: Compiling VBoxMouseFilter - /Opslag/haikuports/app-emulation/virtualbox-guest-additions/work-x86-6.1.26/sources/VirtualBox-6.1.26/src/VBox/Additions/haiku/VBoxMouse/VBoxMouseFilter.cpp
/Opslag/haikuports/app-emulation/virtualbox-guest-additions/work-x86-6.1.26/sources/VirtualBox-6.1.26/src/VBox/Additions/haiku/VBoxMouse/VBoxMouseFilter.cpp:65:1: error: ambiguating new declaration of 'BInputServerFilter* instantiate_input_filter()'
 instantiate_input_filter()
 ^~~~~~~~~~~~~~~~~~~~~~~~
In file included from /Opslag/haikuports/app-emulation/virtualbox-guest-additions/work-x86-6.1.26/sources/VirtualBox-6.1.26/src/VBox/Additions/haiku/VBoxMouse/VBoxMouseFilter.cpp:56:
/Opslag/haikuports/app-emulation/virtualbox-guest-additions/work-x86-6.1.26/sources/VirtualBox-6.1.26/src/VBox/Additions/haiku/VBoxMouse/VBoxMouseFilter.h:55:40: note: old declaration 'BInputServerFilter* instantiate_input_filter()'
 extern "C" _EXPORT BInputServerFilter* instantiate_input_filter();
                                        ^~~~~~~~~~~~~~~~~~~~~~~~
kmk: *** [/packages/kbuild_x86-0.1.9998svn3296-1/.self/data/kBuild/footer
```

**Comment by wattoc** [comment]
Thanks for testing @Begasus, I will get back to this at some point if no one else picks it up in the meantime, just a bit swamped currently.

**Comment by waddlesplash** [comment]
Related: #1768.

**Comment by wattoc** [comment]
@Begasus I've spent some time looking at x86 - think what I said earlier was wrong regarding being able to build it.  I'm actually wondering if it's possible to make this work at all on x86.

I did "fix" the issue with the mouse driver, but upon attempting linking it I think I've hit a problem - I expect kernel drivers on x86 will require being built on GCC2, and the build for the additions expects a minimum of GCC4.  May be possible to use an older version of the source and fix that up but I've no idea how well that'd work going forward.

Anyone else want to advise on this?  I'm no expert on this sort of thing and try to stick with user space.

**Comment by pulkomandy** [comment]
I'd say it's not really useful to waste time trying to get modern code working with gcc2. People should run the 64bit version of Haiku if they want the nice and modern features like this.

**Comment by wattoc** [comment]
I've no problem with that - I find Virtualbox has lots of issues with the 32 bit Haiku for whatever reason anyway.

I've been using the package on my x64 VM continuously now for a while.  There's an issue where the Deskbar will eventually freeze up which will obviously be something in the tray app.  I'll be looking to switch that to some sort of background app if possible but there's obviously something dieing over time there.

Otherwise will try to figure out what's going on with the mouse driver since it doesn't seem to do anything.

If anyone wants to suggest anything re: the shared folders support that'd be helpful as I'm not sure how to progress it further.

File: app-emulation/virtualbox-guest-additions/virtualbox_guest_additions-6.1.26.recipe [hunk 1]
```diff
@@ -0,0 +1,109 @@
...
+	lib:libiconv$secondaryArchSuffix
+	lib:libpng16$secondaryArchSuffix
+	lib:libz$secondaryArchSuffix
+	"
+
+BUILD_REQUIRES="
+	haiku${secondaryArchSuffix}_devel
+	devel:libcurl$secondaryArchSuffix
+	devel:libiconv$secondaryArchSuffix
+	devel:libpng$secondaryArchSuffix
```
Comment (Begasus) [inline app-emulation/virtualbox-guest-additions/virtualbox_guest_additions-6.1.26.recipe:41]: Try to use libpng16 here also as mentioned in the issue https://github.com/haikuports/haikuports/issues/3779 : 
```diff
- 	devel:libpng$secondaryArchSuffix
+ 	devel:libpng16$secondaryArchSuffix
```

**Comment by Begasus** [comment]
What's the status here?
If things are OK on the 64bit side I'm OK with merging here

**Comment by wattoc** [comment]
On 64bit side things generally build/work OK with some issues as above.  I'm continuing to work on it, if you merge now there'll be some more updates.  I guess the balance is between it potentially causing issues for some vs. gaining some useful features.

I'm not sure if any others have been using it on 64bit yet.

File: app-emulation/virtualbox-guest-additions/virtualbox_guest_additions-6.1.26.recipe [hunk 2]
```diff
@@ -0,0 +1,109 @@
...
+	GNU GPL v2
+	CDDL v1
+	"
+REVISION="1"
+SOURCE_URI="http://download.virtualbox.org/virtualbox/$portVersion/VirtualBox-$portVersion.tar.bz2"
+CHECKSUM_SHA256="0212602eea878d6c9fd7f4a3e0182da3e4505f31d25f5539fb8f7b1fbe366195"
+SOURCE_DIR="VirtualBox-$portVersion"
+PATCHES="virtualbox_guest_additions-$portVersion.patchset"
+
+ARCHITECTURES="!x86_gcc2 ?x86 x86_64"
```
Comment (Begasus) [inline app-emulation/virtualbox-guest-additions/virtualbox_guest_additions-6.1.26.recipe:20]: One last request for now :)
Could you change the following? This is after a recent change in Haikuporter: https://github.com/haikuports/haikuporter/commit/da92250cd2ac73492e9b094d1eb561ecca9fd771
```diff
- ARCHITECTURES="!x86_gcc2 ?x86 x86_64"
+ ARCHITECTURES="all !x86_gcc2"
```

Comment (wattoc) [reply app-emulation/virtualbox-guest-additions/virtualbox_guest_additions-6.1.26.recipe:20]: I looked at that change and it confused me.  Does all imply it works on all architectures but the ones specified with !xxx?

It's not really true at all for this package, if anything it should just be ARCHITECTURES="x86_64"?

Comment (Begasus) [reply app-emulation/virtualbox-guest-additions/virtualbox_guest_additions-6.1.26.recipe:20]: @waddlesplash maybe you could elaborate a bit here?

Comment (waddlesplash) [reply app-emulation/virtualbox-guest-additions/virtualbox_guest_additions-6.1.26.recipe:20]: If it really only works on x86_64, then yes, that is the only thing that should be mentioned.

Comment (Begasus) [reply app-emulation/virtualbox-guest-additions/virtualbox_guest_additions-6.1.26.recipe:20]: OK, then maybe `!x86_gcc2` isn't needed in `ARCHITECTURES`?

Comment (wattoc) [reply app-emulation/virtualbox-guest-additions/virtualbox_guest_additions-6.1.26.recipe:20]: My understanding is Virtualbox isn't supported outside of x86 architectures.  Currently it won't be any use on any 32-bit Haiku system either.  So only x86_64 is supported at all.  

If I'm understanding the change made to haikuports correctly, then all is a way to say everything is supported, and then we can explicitly list things that aren't supported as we encounter them or if we know already they are not.

So going forward more generally, it sounds like we're no longer doing the ?arch thing, marking all on everything and then marking everything we know won't be supported with !arch.

Comment (waddlesplash) [reply app-emulation/virtualbox-guest-additions/virtualbox_guest_additions-6.1.26.recipe:20]: Recipes that are highly architecture-specific can avoid using `all` if it makes sense to and use the "old way", and this one probably qualifies.

Comment (waddlesplash) [reply app-emulation/virtualbox-guest-additions/virtualbox_guest_additions-6.1.26.recipe:20]: Essentially, in addition to the "opt-in" model, there is now an "opt-out" model, and the "opt-out" model is now the default.

Comment (wattoc) [reply app-emulation/virtualbox-guest-additions/virtualbox_guest_additions-6.1.26.recipe:20]: Thanks @waddlesplash

**Comment by volo-droid** [comment]
> On 64bit side things generally build/work OK with some issues as above. I'm continuing to work on it, if you merge now there'll be some more updates. I guess the balance is between it potentially causing issues for some vs. gaining some useful features.

@Begasus @wattoc is there any good reason not to merge the PR as is, and continue on the other stuff later on?

**Comment by Begasus** [comment]
Merged ... keep track on the buildmaster, if things can evolve you guys could pick up from here :)

**Pull request merged**